### PR TITLE
[ROCm][CI] Re-enable ROCm CI & make CI changes as per new CI design

### PR DESCRIPTION
**Pull request by akashveramd**
This PR covers most of the features that were part of this PR (unmerged & closed) https://github.com/pytorch/torchtitan/pull/4039.

This PR makes following CI changes-
- Re-enable ROCm CI. 
- Separate badging on README.md to indicate ROCm/CUDA workflow status separately.
- Remove `ciflow/8gpu` label as a trigger for the CUDA-specific workflows.
- Add `ciflow/rocm` label as a trigger for the ROCm-specific workflows.
-- This label can be added manually by PR developers wanting to test on ROCm.
-- This label will also be automatically added by the pytorch bot for any PRs that mention ROCm in the title/description.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -54,6 +54,7 @@ We look forward to your contributions!
 | --- | --- | --- |
 | CPU | - | [![CPU Unit Test](https://github.com/pytorch/torchtitan/actions/workflows/unit_test_cpu.yaml/badge.svg?branch=main)](https://github.com/pytorch/torchtitan/actions/workflows/unit_test_cpu.yaml?query=branch%3Amain) |
 | NVIDIA GPU | [![Integration Tests](https://github.com/pytorch/torchtitan/actions/workflows/integration_test.yaml/badge.svg?branch=main)](https://github.com/pytorch/torchtitan/actions/workflows/integration_test.yaml?query=branch%3Amain) [![H100 Tests](https://github.com/pytorch/torchtitan/actions/workflows/integration_test_h100.yaml/badge.svg?branch=main)](https://github.com/pytorch/torchtitan/actions/workflows/integration_test_h100.yaml?query=branch%3Amain) | [![GPU Unit Tests](https://github.com/pytorch/torchtitan/actions/workflows/unit_test_gpu.yaml/badge.svg?branch=main)](https://github.com/pytorch/torchtitan/actions/workflows/unit_test_gpu.yaml?query=branch%3Amain) |
+| AMD GPU (ROCm) | [![ROCm Integration Tests](https://github.com/pytorch/torchtitan/actions/workflows/integration_test_rocm.yaml/badge.svg?branch=main)](https://github.com/pytorch/torchtitan/actions/workflows/integration_test_rocm.yaml?query=branch%3Amain) | - |
```
Comment (wwwjn) [inline README.md]:
```suggestion
| AMD GPU (ROCm) | [![Integration Tests](https://github.com/pytorch/torchtitan/actions/workflows/integration_test_rocm.yaml/badge.svg?branch=main)](https://github.com/pytorch/torchtitan/actions/workflows/integration_test_rocm.yaml?query=branch%3Amain) | - |
```

**Comment by wwwjn** [comment]
Seems the bot have an error: https://github.com/pytorch/torchtitan/pull/4361#issuecomment-5446286193

**Comment by akashveramd** [comment]
> Seems the bot have an error: [#4361 (comment)](https://github.com/pytorch/torchtitan/pull/4361#issuecomment-5446286193)

@wwwjn: The bot warning was triggered before commit 232d3f1e added ciflow/rocm to .github/pytorch-probot.yml. That commit is now included in this PR, so the label is properly registered, and the warning won't recur once this is merged.

**Comment by rishisinhanj** [comment]
Hi @wwwjn @tianyu-l cc: @akashveramd @jithunnair-amd 

After making some fixes and running the ROCm CI for this PR, we see equality loss issues with golden numerics files that were developed in #4127

From my understanding, every golden numerics file is generated from CUDA A10G runs, and there is no ROCm CI loss file anymore. We wanted to make sure we were aligned on the design you wanted for this moving forward with ROCm in TorchTitan. 

Proposed fix: parameterize golden_numerics_path on arch (e.g. llama3_{gpu_arch}.txt) by threading gpu_arch_type into the .format() call in run_tests.py. We can then add new ROCm-specific golden files generated from real ROCm runs leaving every existing _a10g.txt untouched so CUDA is unaffected. 

If this sounds ok to you, we can add it to this PR. If you have a different idea for this, let us know.

**Comment by tianyu-l** [comment]
I don't think AMD golden loss should be required. The idea is to guard on any accidental numerical change, so verify it on one hardware would be enough.

**Comment by yuankaichen-amd** [comment]
@tianyu-l   we would also need it somewhere to guard on the end-to-end training numerical changes on AMD. Do you have any suggestions?

**Comment by tianyu-l** [comment]
@yuankaichen-amd curious why do you need it? E.g. would there be cases where cuda numerics are unchanged, but rocm numerics are changed?

**Comment by yuankaichen-amd** [comment]
@tianyu-l   Yes. It's likely. Although there are unit tests in pytorch, torchao etc to ensure the AMD kernel correctness. It'd be also valuable to have end-to-end training guards.

**Comment by tianyu-l** [comment]
@yuankaichen-amd OK, then we can host some loss files for rocm. Up to you if you want to host the full spectrum compared with cuda ones. Again, titan devs shouldn't be blocked by them.

**Comment by akashveramd** [comment]
@tianyu-l @wwwjn: The PR is ready for review. Could you please review it?

File: .github/workflows/integration_test.yaml [hunk 2]
```diff
@@ -171,7 +174,10 @@ jobs:
         USE_CPP=0 python -m pip install --pre torchao --index-url ${{ matrix.runner_config['index-url'] }}
 
         if [[ "${{ matrix.runner_config['gpu-arch-type'] }}" == "rocm" ]]; then
-          export HIPBLASLT_TENSILE_LIBPATH="$(python -c 'import os, torch; print(os.path.join(os.path.dirname(torch.__file__), "lib", "hipblaslt", "library"))')"
+          HIPBLASLT_LIB_DIR="$(python -c 'import os, torch; print(os.path.join(os.path.dirname(torch.__file__), "lib", "hipblaslt", "library"))')"
+          if [ -d "${HIPBLASLT_LIB_DIR}" ]; then
+            export HIPBLASLT_TENSILE_LIBPATH="${HIPBLASLT_LIB_DIR}"
```
Comment (jithunnair-amd) [inline .github/workflows/integration_test.yaml:184]: @rishisinhanj Does this mean that the `HIPBLASLT_TENSILE_LIBPATH` env var will be automatically set for ROCm7.14+ to pick up from ROCm wheel installation path?

**Review by jithunnair-amd — APPROVED** [review]
LGTM. @tianyu-l Please let us know if this addresses all the concerns that were raised in previous discussions. We are hoping to re-enable ROCm CI on torchtitan.

File: .github/workflows/integration_test.yaml (hunk 2 above)
Comment (rishisinhanj) [reply .github/workflows/integration_test.yaml:184]: Yes, this guards against if hipblastlt is not in the torch/lib/hipblaslt library. Before this, we hit an error because the old unconditional pointed to a dead path

File: tests/integration_tests/features.py [hunk 3]
```diff
@@ -310,7 +310,9 @@ def build_features_test_list() -> list[OverrideDefinitions]:
             test_descr="SFT ChatDataset integration and numerics test",
             test_name="sft",
             ngpu=2,
-            golden_numerics_path=("tests/assets/losses/{execution_mode}/sft_a10g.txt"),
+            golden_numerics_path=(
+                "tests/assets/losses/{execution_mode}/sft_{gpu_arch}.txt"
```
Comment (wwwjn) [inline tests/integration_tests/features.py]: Will the `gpu_arch` for cuda be "CUDA" or "A10G"? If it's cuda, we should rename the files for CUDA golden loss to avoid breaking cuda test?

File: tests/integration_tests/run_tests.py [hunk 4]
```diff
@@ -244,8 +245,11 @@ def run_single_test(
             # compares them with the mode-specific golden (or exports them).
             assert config_fn is not None and config is not None
             execution_mode = "fake_pg" if use_fake_pg else "real_pg"
+            gpu_arch = "a10g" if gpu_arch_type == "cuda" else "mi350x"
```
Comment (wwwjn) [inline tests/integration_tests/run_tests.py]: This would likely to break if we run the numerics test on H100 / B200 machine? 

I would proposing organize the golden loss folder like "assets/losses/real_pg/mi350x" and "assets/losses/real_pg/a10g" and set the gpu type (eg, mi350x, a10g , h100, b200) in the workflow yaml file?

**Review by wwwjn — COMMENTED** [review]
Thanks for the reenablement, see my following comments

File: tests/integration_tests/run_tests.py (hunk 4 above)
Comment (rishisinhanj) [reply tests/integration_tests/run_tests.py]: Done - we reorganized the loss folder per your suggestion

File: tests/integration_tests/features.py (hunk 3 above)
Comment (rishisinhanj) [reply tests/integration_tests/features.py]: Done - it is a10g set as a literal value in the workflow yaml and CUDA is not affected.

**Comment by rishisinhanj** [comment]
@wwwjn Thanks for the feedback! I've made changes per your request- namely changing the loss file organization and making sure gpu-arch is explicit. The CI remains green. If you have any follow-ups, let us know.

File: run_train.sh [hunk 5]
```diff
@@ -37,6 +37,9 @@ if [ "$COMM_MODE" = "fake_backend" ]; then
     # have to precede the caller-provided arguments.
     NGPU="${NGPU}" LOCAL_RANK=0 python3 -m torchtitan.train --module ${MODULE} --config ${CONFIG} --comm.mode=fake_backend --training.steps 1 "$@"
 else
+    # Clear stale NCCL shm segments leaked by prior abnormal exits.
+    rm -f /dev/shm/nccl-* 2>/dev/null || true
```
Comment (wwwjn) [inline run_train.sh]: Can we remove this line from run_trian.sh? `/dev/shm/nccl` seems not a common folder path for all users

File: README.md [hunk 6]
```diff
@@ -54,6 +54,7 @@ We look forward to your contributions!
...
 | NVIDIA GPU | [![Integration Tests](https://github.com/pytorch/torchtitan/actions/workflows/integration_test.yaml/badge.svg?branch=main)](https://github.com/pytorch/torchtitan/actions/workflows/integration_test.yaml?query=branch%3Amain) [![H100 Tests](https://github.com/pytorch/torchtitan/actions/workflows/integration_test_h100.yaml/badge.svg?branch=main)](https://github.com/pytorch/torchtitan/actions/workflows/integration_test_h100.yaml?query=branch%3Amain) [![B200 Tests](https://github.com/pytorch/torchtitan/actions/workflows/integration_test_b200.yaml/badge.svg?branch=main)](https://github.com/pytorch/torchtitan/actions/workflows/integration_test_b200.yaml?query=branch%3Amain) | [![GPU Unit Tests](https://github.com/pytorch/torchtitan/actions/workflows/unit_test_gpu.yaml/badge.svg?branch=main)](https://github.com/pytorch/torchtitan/actions/workflows/unit_test_gpu.yaml?query=branch%3Amain) |
+| AMD GPU (ROCm) | [![Integration Tests](https://github.com/pytorch/torchtitan/actions/workflows/integration_test_rocm.yaml/badge.svg?branch=main)](https://github.com/pytorch/torchtitan/actions/workflows/integration_test_rocm.yaml?query=branch%3Amain) | - |
```
Comment (wwwjn) [inline README.md]: Could you double check if this link points to the correct badge? When I view file, I see the title to be "Integration Tests - Features (ROCm) "?

**Review by wwwjn — APPROVED** [review]
Thanks for the great work. Please fix my follow comments before merge, and rebase

**Comment by akashveramd** [comment]
@wwwjn: The comments have been resolved, and the PR has been rebased on the latest upstream main.

**Comment by akashveramd** [comment]
@wwwjn @tianyu-l: All checks are now passing. Are we good to merge this PR?
